<a href="https://colab.research.google.com/github/SophistryDude/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Nick's_Copy_of_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving imports-85.data to imports-85.data


In [ ]:
# just getting started.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
df = pd.read_csv('imports-85.data')
print(df.shape)
df.head()

(204, 26)


3    ?  alfa-romero  gas  std   two  ...  9.00  111  5000  21  27  13495
0  3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  16500
1  1    ?  alfa-romero  gas  std   two  ...   9.0  154  5000  19  26  16500
2  2  164         audi  gas  std  four  ...  10.0  102  5500  24  30  13950
3  2  164         audi  gas  std  four  ...   8.0  115  5500  18  22  17450
4  2    ?         audi  gas  std   two  ...   8.5  110  5500  19  25  15250

[5 rows x 26 columns]

In [ ]:
# I pulled the column headers from the names file.
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df = pd.read_csv('imports-85.data', names = column_headers) 
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [ ]:
# Here we are going to get rid of the objects and replace them w/ NaN or float values. 
df = df.replace({'y': 1, 'n': 0, "?": np.NaN})
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [ ]:
### Set the sample size to 10, and use the random state 30 to keep along w/ class. In the future I'll probably use a more poignant variable name, but this keeps w/ the class. 
sample = df.sample(10, random_state=30)
sample.head()

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056

[5 rows x 26 columns]

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

### Hypothesis ###
The fuel efficiency will be approximately equal to or greater than 35mpg on the highway. 

In [ ]:
### Instead of assessing w/ a confidence rating at the beginning, let's analyze the data and say what confidence rating we have at the end. 
stats.ttest_1samp(sample['highway-mpg'], .5, nan_policy='omit')

Ttest_1sampResult(statistic=14.050858301455087, pvalue=1.98756047234675e-07)

## Confidence rating
0.0000001988 is well within the range of 99%. So we can set our confidence rating to 99%. 
## Hypothesis conclusion
1.988 x 10^-7 is less than p-value=.01, therefore the sample size of 10 fails to reject the hypothesis. 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
### YOUR WORK HERE
stats.ttest_1samp(df.sample(100, random_state=30)['highway-mpg'], .5, nan_policy='omit')

Ttest_1sampResult(statistic=45.061541466888855, pvalue=8.84958682057164e-68)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

## Response
When only using 10 vehicles, each vehicle represents 10% of the reported values, and as such is weighted more. When using the 100 vehicles, each vehicle then represents 1% of the reported values. This means that extreme outliers do will not affect the data as much. 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
### YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-08-11 15:12:02--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-08-11 15:12:03 (286 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [ ]:
column_headers = ['class name', 'handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution', 'physician-free-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile', 'immigration', 'synfuels-corporation-cutback', 'education-spending', 'superfund-right-to-sue', 'crime', 'duty-free-exports', 'south-africa-act']
house_votes = pd.read_csv('house-votes-84.data', names= column_headers)
print(house_votes.shape)
house_votes.head()


(435, 17)


class name handicapped-infants  ... duty-free-exports south-africa-act
0  republican                   n  ...                 n                y
1  republican                   n  ...                 n                ?
2    democrat                   ?  ...                 n                n
3    democrat                   n  ...                 n                y
4    democrat                   y  ...                 y                y

[5 rows x 17 columns]

In [ ]:
# Same as the 1 sample ttest. 
house_votes = house_votes.replace({'y':1, 'n':0, '?': np.NaN})
house_votes.head()

class name  handicapped-infants  ...  duty-free-exports  south-africa-act
0  republican                  0.0  ...                0.0               1.0
1  republican                  0.0  ...                0.0               NaN
2    democrat                  NaN  ...                0.0               0.0
3    democrat                  0.0  ...                0.0               1.0
4    democrat                  1.0  ...                1.0               1.0

[5 rows x 17 columns]

In [ ]:
# For the two samples we needed to isolate based off of the party type. While this solution wouldn't work with a large variation of objects, I should be able to edit the .str to allow for sufficient amount of chars
rep = house_votes[house_votes['class name'].str[0] == 'r']
rep.head()
#edit: it turns out that I don't need the .str function at all. a cleaner version of the code is: 
  # rep = df[df['party'] == 'republican]

class name  handicapped-infants  ...  duty-free-exports  south-africa-act
0   republican                  0.0  ...                0.0               1.0
1   republican                  0.0  ...                0.0               NaN
7   republican                  0.0  ...                NaN               1.0
8   republican                  0.0  ...                0.0               1.0
10  republican                  0.0  ...                0.0               0.0

[5 rows x 17 columns]

In [ ]:
dem = house_votes[house_votes['class name'].str[0] == 'd']
dem.head()

class name  handicapped-infants  ...  duty-free-exports  south-africa-act
2   democrat                  NaN  ...                0.0               0.0
3   democrat                  0.0  ...                0.0               1.0
4   democrat                  1.0  ...                1.0               1.0
5   democrat                  0.0  ...                1.0               1.0
6   democrat                  0.0  ...                1.0               1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

# Hypothesis 
-------------------------------------------------------------------
## **Religious groups in schools**
### Null hypothesis:
They support the "Religous groups in schools" bill equally. 
###  Alternate hypothesis: 
Republicans support the "religious groups in schools" with a p values < 0.01

## **Handicapped-infants**
### Null hypothesis: 
They support the "handicapped-infants" bill equally
### Alternate hypothesis: 
Democrats support the "handicapped-infants" with a p value < 0.01
## **Immigration**
### Null hypothesis
They support the immigration bill equally
### The republicans support the immigration bill more w/ a p value of > 0.1

In [ ]:
stats.ttest_ind(dem['religious-groups-in-schools'], rep['religious-groups-in-schools'], nan_policy='omit')

Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

Null hypothesis incorrect. Alternate hypothesis is correct. This bill is skewed towards republicans with a p value of 0.00000000000000000002394.


In [ ]:
### YOUR WORK HERE
stats.ttest_ind(dem['handicapped-infants'], rep['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

Null hypothesis incorrect. Alternate hypothesis is correct. This bill is skewed towards democrats with a p value of 0.000000000000000001613

In [ ]:
stats.ttest_ind(dem['immigration'], rep['immigration'], nan_policy='omit')

Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

Null hypothesis is correct. This bill has a pvalue of 0.08, which is within the parameters set at confidence rating 90%. It's noteworthy that the bill is slightly more favorable by republicans. 

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
# You can use a for/in/if loop to replace the specific column tested above, and test all the columns. 
test = {}
for column_headers in house_votes:
  # '' lets us search all columns, if we wanted we could search a specific column with that as well. 
  if column_headers != 'class name':
    test[column_headers] = stats.ttest_ind(rep[column_headers], dem[column_headers])

test

{'adoption-of-the-budget-resolution': Ttest_indResult(statistic=nan, pvalue=nan),
 'aid-to-nicaraguan-contras': Ttest_indResult(statistic=nan, pvalue=nan),
 'anti-satellite-test-ban': Ttest_indResult(statistic=nan, pvalue=nan),
 'crime': Ttest_indResult(statistic=nan, pvalue=nan),
 'duty-free-exports': Ttest_indResult(statistic=nan, pvalue=nan),
 'education-spending': Ttest_indResult(statistic=nan, pvalue=nan),
 'el-salvador-aid': Ttest_indResult(statistic=nan, pvalue=nan),
 'handicapped-infants': Ttest_indResult(statistic=nan, pvalue=nan),
 'immigration': Ttest_indResult(statistic=nan, pvalue=nan),
 'mx-missile': Ttest_indResult(statistic=nan, pvalue=nan),
 'physician-free-freeze': Ttest_indResult(statistic=nan, pvalue=nan),
 'religious-groups-in-schools': Ttest_indResult(statistic=nan, pvalue=nan),
 'south-africa-act': Ttest_indResult(statistic=nan, pvalue=nan),
 'superfund-right-to-sue': Ttest_indResult(statistic=nan, pvalue=nan),
 'synfuels-corporation-cutback': Ttest_indResult(sta

In [ ]:
# When the loop function operates, it returns nan values. So we need to omit the nans. 
test = {}
for column_headers in house_votes:
  if column_headers != 'class name':
    #I forgot the nan_policy='omit'!
    test[column_headers] = stats.ttest_ind(rep[column_headers], dem[column_headers], nan_policy='omit')

test

{'adoption-of-the-budget-resolution': Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77),
 'aid-to-nicaraguan-contras': Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54),
 'anti-satellite-test-ban': Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31),
 'crime': Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47),
 'duty-free-exports': Ttest_indResult(statistic=-12.853146132542978, pvalue=5.997697174347365e-32),
 'education-spending': Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64),
 'el-salvador-aid': Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68),
 'handicapped-infants': Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18),
 'immigration': Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066),
 'mx-missile': Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47),
 'p



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





![One Sample T Statistic](https://www.statisticshowto.com/wp-content/uploads/2015/09/t-score.jpg)


In [ ]:
### YOUR WORK HERE
#stats.ttest_1samp(df.sample(100, random_state=30)['highway-mpg'], .5, nan_policy='omit')
root_n = 

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE